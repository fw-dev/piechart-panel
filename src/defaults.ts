import { IPanelOptions, IChartConfig, IChartData, IHighlight } from 'types';

export const defaultPanelOptions: IPanelOptions = {
  title: 'Title',
  chartType: 'pie',
  legendEnabled: true,
  legendPosition: 'left',
  legendAlign: 'center',
  legendBoxWidth: '40',
  legendFontSize: '12',
  legendUsePointStyle: false,
  linkEnabled: false,
  linkUrl: '',
  linkTargetBlank: true,
  highlightEnabled: true,
  highlightValue: 'percentage',
  selectedHighlight: '',
  nullPointMode: 'connected',
  valueName: 'current',
  aliasColors: {},
  cutoutPercentage: '80',
};

export const defaultChartConfig: IChartConfig = {
  type: 'pie',
  options: {
    cutoutPercentage: 80,
    responsive: true,
    aspectRatio: 1,
    maintainAspectRatio: false,
    animation: {
      animateRotate: false,
    },
  },
};

export const defaultChartData: IChartData = {
  labels: [],
  datasets: [{
    data: [],
    backgroundColor: [],
  }]
};

export const defaultHighlight: IHighlight = {
  value: '',
  label: '',
};


export const chartOptions = {
  position: [
    { label: 'top', value: 'top' },
    { label: 'left', value: 'left' },
    { label: 'bottom', value: 'bottom' },
    { label: 'right', value: 'right' },
  ],
  align: [
    { label: 'start', value: 'start' },
    { label: 'center', value: 'center' },
    { label: 'end', value: 'end' },
  ],
  chartType: [
    { label: 'Pie', value: 'pie' },
    { label: 'Doughnut', value: 'doughnut' },
  ],
  highlight: [{ label: 'test', value: 'test' }, { label: 'hallo', value: 'hallo' }],
  highlightValue: [{ label: 'Percentage', value: 'percentage' }, { label: 'Number', value: 'number' }],
};

export const colors = ['#96D98D', '#8AB8FF', '#EB4034', '#EBDC34'];