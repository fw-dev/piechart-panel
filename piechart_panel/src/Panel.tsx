import React from 'react';
// @ts-ignore
import Chart from 'chart.js';
// @ts-ignore
import TimeSeries from './TimeSeries';

import { State, Props } from './types';
import { defaultChartConfig, initialState, colors, defaultHighlight } from './defaults';
import 'css/filewave-piechart-panel.css';

export class Panel extends React.Component<Props, State> {
  panelId: number | undefined = this.props.data.request && this.props.data.request.panelId;
  chart: any;
  state = initialState;

  componentDidMount() {
    this.handleDataFormatting(this.props.data);
  }

  componentDidUpdate(prevProps: Props, prevState: State) {
    const { options, data } = this.props;
    const { highlight, chartData } = this.state;

    if (options !== prevProps.options) {
      this.updateChart();
    }

    if (highlight.label !== prevState.highlight.label || options.highlightEnabled !== prevProps.options.highlightEnabled) {
      this.drawChart();
    }

    if (data !== prevProps.data || options.aliasColors !== prevProps.options.aliasColors) {
      this.handleDataFormatting(data);
    }

    if (chartData !== prevState.chartData) {
      this.updateHighlight();
    }
  }

  updateHighlight = () => {
    const { selectedHighlight } = this.props.options;
    const { highlightData } = this.state;
    const highlight = highlightData.find(highlight => highlight.label === selectedHighlight) || defaultHighlight;
    this.setState({ highlight });
  };

  handleDataFormatting = (data: any) => {
    const timeSeries = data.series.map((serie: any) => this.createTimeSeries(serie));
    const chartData = this.formatChartData(timeSeries, data.series);
    const highlightData = this.formatHighlightData(timeSeries);
    this.setState({
      chartData,
      highlightData,
    });
  };

  formatHighlightData = (timeSeries: any) => {
    const { options } = this.props;
    const total = timeSeries.reduce((x: number, y: any) => x + y.stats.total, 0);
    const { highlightValue } = this.props.options;
    const highlightData = timeSeries.map((serie: any) => {
      const percentage = `${(serie.stats[options.valueName] / (total / 100) || 0).toFixed()}%`;
      const value = serie.stats[options.valueName];
      return {
        label: serie.label,
        value: highlightValue === 'percentage' ? percentage : value,
      };
    });

    return highlightData;
  };

  formatChartData = (timeSeries: any, series: any) => {
    const { valueName, aliasColors } = this.props.options;
    const chartData = {
      labels: timeSeries.map((serie: any) => serie.alias),
      datasets: [
        {
          data: timeSeries.map((serie: any) => serie.stats[valueName]),
          backgroundColor: timeSeries.map((serie: any, i: number) => aliasColors[serie.alias] || colors[i]),
          metadata: series.map((serie: any) => serie.fields.find((field: any) => field.labels).labels),
        },
      ],
    };

    return chartData;
  };

  createTimeSeries = (serie: any) => {
    const timeSeries = new TimeSeries({
      datapoints: [serie.fields.map((field: any) => field.values.buffer[0])],
      alias: serie.name,
      target: serie.name,
    });
    timeSeries.flotpairs = timeSeries.getFlotPairs(this.props.options.nullPointMode);

    return timeSeries;
  };

  updateChartSettings = () => {
    const { options, data } = this.props;
    return {
      ...defaultChartConfig,
      data: this.state.chartData,
      type: options.chartType,
      plugins: [
        {
          afterDraw: () => {
            if (!data.series.length) {
              return this.drawDataUnavailableMessage();
            }
            if (options.highlightEnabled && options.chartType === 'doughnut') {
              return this.drawHighlight();
            }
          },
        },
      ],
      options: {
        ...defaultChartConfig.options,
        cutoutPercentage: options.chartType === 'doughnut' ? parseInt(options.cutoutPercentage, 0) : 0,
        onClick: this.handleClick,
        legend: {
          display: options.legendEnabled,
          position: options.legendPosition,
          align: options.legendAlign,
          labels: {
            boxWidth: parseInt(options.legendBoxWidth, 0),
            fontSize: parseInt(options.legendFontSize, 0),
            usePointStyle: options.legendUsePointStyle,
          },
        },
      },
    };
  };

  generateUrl = (target: any) => {
    const { linkUrl } = this.props.options;
    let url = linkUrl;
    const variableRegex = /(\${__)(.*?)(})/g; // Match anything that's written between ${__...}
    const queryVariables = Object.keys(target.metadata).map((variable: string) => ({ name: variable, value: target.metadata[variable] }));
    const vars = [...queryVariables];
    // @ts-ignore, need to look into fixing "Object is possibly null" here
    const matchedVars = linkUrl.match(variableRegex).map((variable: string) => {
      const matchedVariable = vars.find((item: any) => item.name === variable.substring(4, variable.length - 1)) || { value: '', name: '' };
      return {
        name: variable,
        value: variable.includes('targetLabel') ? target.label : variable.includes('targetValue') ? target.value : matchedVariable.value || variable,
      };
    });
    matchedVars.forEach((match: any) => (url = url.replace(match.name, match.value)));

    return url;
  };

  handleClick = (_event: Event, targets: any) => {
    const { options } = this.props;
    if (!options.linkEnabled || !targets.length) {
      return;
    }

    const { labels, datasets } = this.chart.data;
    const targetIndex = targets[0]._index;
    const url = this.generateUrl({
      label: labels[targetIndex],
      value: datasets[0].data[targetIndex],
      metadata: datasets[0].metadata[targetIndex],
    });

    return window.open(url, options.linkTargetBlank ? '_blank' : 'currentWindow');
  };

  updateChart = () => {
    const { options } = this.props;
    if (this.chart.config.type !== options.chartType) {
      return this.drawChart();
    }
    if (options.highlightEnabled && options.selectedHighlight) {
      this.updateHighlight();
    }
    this.chart.options = this.updateChartSettings().options;
    this.chart.update({ duration: 0 });
  };

  drawDataUnavailableMessage = () => {
    const {
      ctx,
      chartArea: { right, bottom },
    } = this.chart.chart;
    const xPos = right;
    const yPos = bottom / 2;
    this.drawText(ctx, this.props.options.dataUnavailableMessage, xPos, yPos, 18);
  };

  drawHighlight = () => {
    const {
      ctx,
      width,
      height,
      chartArea: { left, right, top, bottom },
    } = this.chart.chart;
    const { position } = this.chart.legend;
    const { label, value } = this.state.highlight;
    const xPos = Math.round(position === 'left' ? width + left : right);
    const yPos = (position === 'top' ? height + top : bottom) / 2;
    this.drawText(ctx, value, xPos, yPos - 5, 32);
    this.drawText(ctx, label, xPos, yPos + 25, 18);
  };

  drawText = (ctx: any, value: any, xPos: number, yPos: number, fontSize: number) => {
    ctx.restore();
    ctx.textBaseline = 'middle';
    ctx.font = `${fontSize}px sans-serif`;
    ctx.fillStyle = '#BABABA';
    xPos = (xPos - ctx.measureText(value).width) / 2;
    ctx.fillText(value, xPos, yPos);
    ctx.save();
  };

  drawChart = () => {
    const canvas: HTMLElement | HTMLCanvasElement | null = document.getElementById(`fw-piechart-${this.panelId}`);
    if (this.chart && this.chart.id >= 0) {
      // If a chart exists, destroy it, otherwise it will add another one to the canvas
      this.chart.destroy();
    }

    if (canvas instanceof HTMLCanvasElement) {
      this.chart = new Chart(canvas.getContext('2d'), { ...this.updateChartSettings() });
    }
  };

  render() {
    const { width, height } = this.props;
    return (
      <div className="fw-piechart" style={{ width, height }}>
        <canvas id={`fw-piechart-${this.panelId}`} />
      </div>
    );
  }
}
