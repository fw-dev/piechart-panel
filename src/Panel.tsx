import React from 'react';
import { PanelProps } from '@grafana/data';
// @ts-ignore
import Chart from 'chart.js';
// @ts-ignore
import TimeSeries from './TimeSeries';

import { PanelOptions, State } from './types';
import { defaultChartConfig, defaultChartData, defaultHighlight, colors } from './defaults';
import 'css/filewave-piechart-panel.css';

interface Props extends PanelProps<PanelOptions> {}

export class Panel extends React.PureComponent<Props, State> {
  private chart: any;

  constructor(props: Props) {
    super(props);
    this.state = {
      chartData: defaultChartData,
      highlight: defaultHighlight,
    };
    this.chart = React.createRef();
  }

  componentDidMount() {
    this.handleDataFormatting(this.props.data);
  }

  componentDidUpdate(prevProps: Props, prevState: State) {
    const { options, data } = this.props;
    const { highlight, chartData } = this.state;

    if (options !== prevProps.options) {
      this.updateChart();
    }

    if (highlight !== prevState.highlight && options.highlightEnabled) {
      this.drawChart();
    }

    if (data !== prevProps.data || options.aliasColors !== prevProps.options.aliasColors) {
      this.handleDataFormatting(data);
    }

    if (chartData !== prevState.chartData) {
      this.updateHighlight();
    }
  }

  updateHighlight = () => {
    const { selectedHighlight, highlightValue } = this.props.options;
    const { chartData } = this.state;
    const chartValues = chartData.datasets[0].data || [];
    const highlightIndex = chartData.labels.indexOf(selectedHighlight);
    const total = chartValues.reduce((x: number, y: number) => x + y, 0);
    const value = chartValues[highlightIndex] || 0;
    const percentage = `${(value / (total / 100) || 0).toFixed()}%`;

    this.setState({
      highlight: {
        value: highlightValue === 'percentage' ? percentage : value.toString(),
        label: selectedHighlight,
      },
    });
  };

  handleDataFormatting = (data: any) => {
    const timeSeries = data.series.map((serie: any) => this.createTimeSeries(serie));
    const chartData = this.formatChartData(timeSeries, data.series);
    this.setState({ chartData });
    this.drawChart();
  };

  formatChartData = (timeSeries: any, series: any) => {
    const { options } = this.props;
    const { aliasColors } = options;
    const chartData = {
      labels: timeSeries.map((serie: any) => serie.alias),
      datasets: [
        {
          data: timeSeries.map((serie: any) => serie.stats[options.valueName]),
          backgroundColor: timeSeries.map((serie: any, i: number) => aliasColors[serie.alias] || colors[i]),
          metadata: series.map((serie: any) => serie.fields.find((field: any) => field.labels).labels),
        },
      ],
    };

    return chartData;
  };

  createTimeSeries = (serie: any) => {
    const timeSeries = new TimeSeries({
      datapoints: [serie.fields.map((field: any) => field.values.buffer[0])],
      alias: serie.name,
      target: serie.name,
    });
    timeSeries.flotpairs = timeSeries.getFlotPairs(this.props.options.nullPointMode);

    return timeSeries;
  };

  updateChartSettings = () => {
    const { options, data } = this.props;
    return {
      ...defaultChartConfig,
      data: this.state.chartData,
      type: options.chartType,
      plugins: [
        {
          afterDraw: () => {
            if (!data.series.length) {
              return this.drawDataUnavailableMessage();
            }
            if (options.highlightEnabled && options.chartType === 'doughnut') {
              return this.drawHighlight();
            }
          },
        },
      ],
      options: {
        ...defaultChartConfig.options,
        cutoutPercentage: options.chartType === 'doughnut' ? parseInt(options.cutoutPercentage, 0) : 0,
        onClick: this.handleClick,
        legend: {
          display: options.legendEnabled,
          position: options.legendPosition,
          align: options.legendAlign,
          labels: {
            boxWidth: parseInt(options.legendBoxWidth, 0),
            fontSize: parseInt(options.legendFontSize, 0),
            usePointStyle: options.legendUsePointStyle,
          },
        },
      },
    };
  };

  generateUrl = (target: any) => {
    const { linkUrl } = this.props.options;
    let url = linkUrl;
    const variableRegex = /(\${__)(.*?)(})/g; // Match anything that's written between ${__...}
    const queryVariables = Object.keys(target.metadata).map((variable: string) => ({ name: variable, value: target.metadata[variable] }));
    const vars = [...queryVariables];
    // @ts-ignore, need to look into fixing "Object is possibly null" here
    const matchedVars = linkUrl.match(variableRegex).map((variable: string) => {
      const matchedVariable = vars.find((item: any) => item.name === variable.substring(4, variable.length - 1)) || { value: '', name: '' };
      return {
        name: variable,
        value: variable.includes('targetLabel') ? target.label : variable.includes('targetValue') ? target.value : matchedVariable.value || variable,
      };
    });
    matchedVars.forEach((match: any) => (url = url.replace(match.name, match.value)));

    return url;
  };

  handleClick = (_event: any, targets: any) => {
    const { options } = this.props;
    if (!options.linkEnabled || !targets.length) {
      return;
    }

    const { labels, datasets } = this.chart.data;
    const targetIndex = targets[0]._index;
    const url = this.generateUrl({
      label: labels[targetIndex],
      value: datasets[0].data[targetIndex],
      metadata: datasets[0].metadata[targetIndex],
    });

    return window.open(url, options.linkTargetBlank ? '_blank' : 'currentWindow');
  };

  updateChart = () => {
    const { options } = this.props;
    if (this.chart.config.type !== options.chartType) {
      return this.drawChart();
    }
    if (options.highlightEnabled && options.selectedHighlight) {
      this.updateHighlight();
    }
    this.chart.options = this.updateChartSettings().options;
    this.chart.update({ duration: 0 });
  };

  drawDataUnavailableMessage = () => {
    const { ctx } = this.chart.chart;
    const { right, bottom } = this.chart.chart.chartArea;
    const xPos = right;
    const yPos = bottom / 2;
    this.drawText(ctx, this.props.options.dataUnavailableMessage, xPos, yPos, 18);
  };

  drawHighlight = () => {
    const { ctx } = this.chart.chart;
    const { position } = this.chart.legend;
    const { left, right, top, bottom } = this.chart.chart.chartArea;
    const { width, height } = this.chart.chart;
    const { label, value } = this.state.highlight;
    const xPos = Math.round(position === 'left' ? width + left : right);
    const yPos = (position === 'top' ? height + top : bottom) / 2;
    this.drawText(ctx, value, xPos, yPos - 5, 32);
    this.drawText(ctx, label, xPos, yPos + 25, 18);
  };

  drawText = (ctx: any, value: string, xPos: number, yPos: number, fontSize: number) => {
    ctx.restore();
    ctx.textBaseline = 'middle';
    ctx.font = `${fontSize}px sans-serif`;
    ctx.fillStyle = '#BABABA';
    xPos = (xPos - ctx.measureText(value).width) / 2;
    ctx.fillText(value, xPos, yPos);
    ctx.save();
  };

  drawChart = () => {
    if (this.chart.id >= 0) {
      // Destroy the chart, otherwise it will add another one to the canvas
      this.chart.destroy();
    }
    this.chart = new Chart(this.chart.current, { ...this.updateChartSettings() });
  };

  render() {
    const { width, height } = this.props;
    return (
      <div className="fw-piechart" style={{ width, height }}>
        <canvas ref={this.chart} />
      </div>
    );
  }
}
