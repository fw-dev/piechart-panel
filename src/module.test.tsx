import React from 'react';
import 'jest-canvas-mock';
// @ts-ignore
import { configure, mount } from 'enzyme';
// @ts-ignore
import Adapter from 'enzyme-adapter-react-16';
import { Panel } from './Panel';
import { defaultPanelOptions } from './defaults';

configure({ adapter: new Adapter() });

jest.mock('TimeSeries', () => {
  return jest.fn().mockImplementation(() => ({
    getFlotPairs: jest.fn(),
    stats: {},
  }));
});

const defaultProps = {
  id: 1,
  data: {
    state: 'Done',
    series: [
      {
        fields: [
          {
            calcs: null,
            config: { unit: undefined },
            labels: { __name__: 'go_memstats_alloc_bytes_total', instance: 'localhost:20448', job: 'grafana' },
            name: 'grafana',
            type: 'number',
            values: {
              buffer: [1078024840],
            },
          },
        ],
        length: 1,
        meta: undefined,
        name: 'grafana',
        refId: 'A',
      },
    ],
    timeRange: '',
  },
  timeRange: '',
  timeZone: '',
  options: {
    ...defaultPanelOptions,
    aliasColors: {},
  },
  onOptionsChange: jest.fn(),
  renderCounter: null,
  transparent: false,
  width: 500,
  height: 500,
  replaceVariables: jest.fn(),
  onChangeTime: jest.fn(),
  onChangeTimeRange: jest.fn(),
};

describe('Piechart Panel', () => {
  describe('when doing a basic render', () => {
    // Arrange
    const component = <Panel {...defaultProps} />;
    // Act
    const mounted = mount(component);

    // Assert
    it('should render correctly', () => {
      expect(mounted).toMatchSnapshot();
    });
  });
});
