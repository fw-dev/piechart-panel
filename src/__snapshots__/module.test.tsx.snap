// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Piechart Panel when doing a basic render with data should render correctly 1`] = `
<Panel
  data={
    Object {
      "series": Array [
        Object {
          "fields": Array [
            Object {
              "calcs": null,
              "config": Object {
                "unit": undefined,
              },
              "labels": Object {
                "__name__": "go_memstats_alloc_bytes_total",
                "instance": "localhost:20448",
                "job": "grafana",
              },
              "name": "grafana",
              "type": "number",
              "values": Object {
                "buffer": Array [
                  1078024840,
                ],
              },
            },
          ],
          "length": 1,
          "meta": undefined,
          "name": "grafana",
          "refId": "A",
        },
      ],
      "state": "Done",
      "timeRange": "",
    }
  }
  height={500}
  id={1}
  onChangeTime={[MockFunction]}
  onChangeTimeRange={[MockFunction]}
  onOptionsChange={[MockFunction]}
  options={
    Object {
      "aliasColors": Object {},
      "chartType": "pie",
      "cutoutPercentage": "80",
      "dataUnavailableMessage": "Data unavailable",
      "highlightEnabled": true,
      "highlightValue": "percentage",
      "legendAlign": "center",
      "legendBoxWidth": "40",
      "legendEnabled": true,
      "legendFontSize": "12",
      "legendPosition": "left",
      "legendUsePointStyle": false,
      "linkEnabled": false,
      "linkTargetBlank": true,
      "linkUrl": "",
      "nullPointMode": "connected",
      "selectedHighlight": "",
      "valueName": "current",
    }
  }
  renderCounter={null}
  replaceVariables={[MockFunction]}
  timeRange=""
  timeZone=""
  transparent={false}
  width={500}
>
  <div
    className="fw-piechart"
    style={
      Object {
        "height": 500,
        "width": 500,
      }
    }
  >
    <canvas />
  </div>
</Panel>
`;
